packages/react-dom/src/__tests__/validateDOMNesting-test.internal.js

```diff
@@ -9,176 +9,67 @@
-  for (var i = 0; i < stack.length; i++) {
-    if (!validateDOMNesting.isTagValidInContext(stack[i], ancestorInfo)) {
-      return false;
-    }
-    ancestorInfo = validateDOMNesting.updatedAncestorInfo(
-      ancestorInfo,
-      stack[i],
-      null,
-    );
+  let element = null;
+  const warningCount = shouldWarn ? 1 : 0;
+  tags = [...tags];
+  while (tags.length) {
+    element = React.createElement(tags.pop(), null, element);
   }
-  return true;
+
+  spyOnDev(console, 'error');
+  const container = document.createElement('div');
+  ReactDOM.render(element, container);
+  expect(console.error.calls.count()).toEqual(warningCount);
 }
 
 describe('validateDOMNesting', () => {
   beforeEach(() => {
     jest.resetModules();
-
-    // TODO: can we express this test with only public API?
-    validateDOMNesting = require('../client/validateDOMNesting').default;
-  });
-
-  it('allows any tag with no context', () => {
-    if (__DEV__) {
-      // With renderToString (for example), we don't know where we're mounting the
-      // tag so we must err on the side of leniency.
-      var allTags = [].concat(specialTags, formattingTags, ['mysterytag']);
-      allTags.forEach(function(tag) {
-        expect(validateDOMNesting.isTagValidInContext(tag, null)).toBe(true);
-      });
-    }
+    React = require('react');
+    ReactDOM = require('react-dom');
   });
 
   it('allows valid nestings', () => {
     if (__DEV__) {
-      expect(isTagStackValid(['table', 'tbody', 'tr', 'td', 'b'])).toBe(true);
-      expect(isTagStackValid(['body', 'datalist', 'option'])).toBe(true);
-      expect(isTagStackValid(['div', 'a', 'object', 'a'])).toBe(true);
-      expect(isTagStackValid(['div', 'p', 'button', 'p'])).toBe(true);
-      expect(isTagStackValid(['p', 'svg', 'foreignObject', 'p'])).toBe(true);
-      expect(isTagStackValid(['html', 'body', 'div'])).toBe(true);
+      expectInvalidNestingWarning(false, ['table', 'tbody', 'tr', 'td', 'b']);
+      expectInvalidNestingWarning(false, ['div', 'a', 'object', 'a']);
+      expectInvalidNestingWarning(false, ['div', 'p', 'button', 'p']);
+      expectInvalidNestingWarning(false, ['p', 'svg', 'foreignObject', 'p']);
 
       // Invalid, but not changed by browser parsing so we allow them
-      expect(isTagStackValid(['div', 'ul', 'ul', 'li'])).toBe(true);
-      expect(isTagStackValid(['div', 'label', 'div'])).toBe(true);
-      expect(isTagStackValid(['div', 'ul', 'li', 'section', 'li'])).toBe(true);
-      expect(isTagStackValid(['div', 'ul', 'li', 'dd', 'li'])).toBe(true);
+      expectInvalidNestingWarning(false, ['div', 'ul', 'ul', 'li']);
+      expectInvalidNestingWarning(false, ['div', 'label', 'div']);
+      expectInvalidNestingWarning(false, ['div', 'ul', 'li', 'section', 'li']);
+      expectInvalidNestingWarning(false, ['div', 'ul', 'li', 'dd', 'li']);
+
+      // TODO: Previously used test scenarios which will fail now because
+      // root element is div. Remove them?
```

Comment:
No. Let's change the code in `expectInvalidNestingWarning` to treat the first item in the array as the *root element itself*.
For example, for `['div', 'label', 'div']` have it render `<label><div /></label>` into a DOM `div`.